### PR TITLE
Initial docs for gimlet-on-a-bench

### PR DESCRIPTION
**Pull request by leftwo**
Starter doc for how to run a gimlet on a bench.

Lets make this better!

### CODE REVIEW

File: docs/how-to-run-bench-gimlet.adoc [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+export SOFTNPU_MAC=a8:e1:de:01:70:1d
+
+opt/oxide/softnpu/stuff/scadm \
+  --server /opt/oxide/softnpu/stuff/server \
+  --client /opt/oxide/softnpu/stuff/client \
+  standalone \
+  add-proxy-arp \
+  $IP_POOL_START \
+  $IP_POOL_END \
+  $SOFTNPU_MAC
```
Comment (luqmana) [inline docs/how-to-run-bench-gimlet.adoc:232]: So in the case where an entire subnet is allocated to us and the upstream router is configured to route said subnet to us, we don't actually need to configure proxy arp (*) for the ip pools.

(*) that is except for `uplink_ip` itself (on a real sidecar we add the address on the qsfp0 link in the switch zone and so ARPs for it get handled normally).

File: docs/how-to-run-bench-gimlet.adoc [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+----
+./target/release/omicron-package --target sn21 package
+----
+
+Once completed, copy these files over to the gimlet:
+[source,text]
+----
+rsync --delete -Paz tools out \
+  ./target/release/omicron-package package-manifest.toml \
+  --exclude /out/downloads/netstack.* \
```
[the remark this answers is not in the document]
Comment (luqmana) [reply docs/how-to-run-bench-gimlet.adoc:150]: This is the [script](https://gist.github.com/bc3fb26a5cbb64a863af3693e41d85ba) I use for reference

**Review by lifning — COMMENTED** [review]
noted a few typo-tier errors i found while following this.

trying this on our local bench-gimlet, sled-agent spins forever on "Waiting for boot disk M.2..." -- did i (or the doc) miss something obvious?

it might also be worth a brief explanation of why omicron-package is using 'non-gimlet' here rather than 'gimlet-standalone'

File: docs/how-to-run-bench-gimlet.adoc [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+
+You'll also need to configure Proxy ARP.
+The ip pool start and ip pool end also come from your config-rss.toml
+`IP_POOL_START` is [rack_network_config] infra_ip_first
+`IP_POOL_END`   is [rack_network_config] infra_ip_last
+[source,console]
+----
+export SOFTNPU_MAC=a8:e1:de:01:70:1d
+
+opt/oxide/softnpu/stuff/scadm \
```
[the remark this answers is not in the document]
Comment (leftwo) [reply docs/how-to-run-bench-gimlet.adoc]: Yes, thanks for helping to clean this up.

**Review by leftwo — COMMENTED** [review]
Thanks @lifning for the updates here.  I've been buried in prep for FCS and this document was neglected while the underlying software kept moving forward.

Given that I've gotten another bench gimlet working with just a single K.2, I'm going to come back through this document and clean it up a bit more.

**Comment by iliana** [comment]
Something that came up while I was talking about the North Seattle Bench Gimlet with @lifning: I recently learned that the create_virtual_hardware.sh script does not create virtual vdevs while following this documentation, which appears to be what we want, but this behavior is accidental:

https://github.com/oxidecomputer/omicron/blob/09e9d8db09388335ed866471937d4a98f5bd44e8/tools/create_virtual_hardware.sh#L37-L59

i.e. if $OMICRON_TOP/smf/sled-agent/non-gimlet/config.toml does not exist, which it doesn't if you are following these instructions, the grep returns nothing (and pipefail is not set), so vdevs are _silently_ not created.

I feel like this should somehow be made more deliberate but I'm not sure how best to do that.

**Comment by davepacheco** [comment]
> I feel like this should somehow be made more deliberate but I'm not sure how best to do that.

Could it make sense to move more of this script into sled agent as configurable?  i.e., you'd configure sled agent with either "use whatever zpools you find" (what it does today) or "create N file-backed zpools using files in this path, etc."  I feel like doing that with more of these things would streamline the standard redeploy flow and also make it a lot less error-prone.  But I haven't looked into what would be required to do it.

**Comment by leftwo** [comment]
Ah, @iliana  that explains some questions I had that I never looked into.  When I was first trying to figure this out I wondered why the gimlet ended up using the real disks and not what I thought the create_virtual_hardware.sh was "creating".
It worked like I wanted, so I never went back to see why.

File: docs/how-to-run-bench-gimlet.adoc [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+
+You should update `-N` (name of your image) and `-o` (output location of build
+files) for your setup and needs.
+
+[source,text]
+----
+ds=$(TZ=US/Pacific date +%Y%m%d-%H%M%S)
+./helios-build experiment-image \
+  -N "gimlet-sn21-$ds" \
+  -p helios-netdev=https://pkg.oxide.computer/helios-netdev \
```
Comment (citrus-it) [inline docs/how-to-run-bench-gimlet.adoc:75]: For heliosv2:
```suggestion
  -p helios-dev=https://pkg.oxide.computer/helios/2/dev \
```

File: docs/how-to-run-bench-gimlet.adoc [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+
+Once that finishes, you are ready to build your helios image.  Below is an
+example command used to build the image for gimlet-sn21.
+
+The optional `-P` argument arranges for anything in that directory to be laid
+on top of the base image, which we presently use to, say, add in the
+sled-agent SMF manifest and binaries etc.  For example if you `mkdir /tmp/proto`
+and create `/tmp/proto/usr/bin/blah` then passing the `-P /tmp/proto` to your
+build command would make `/tmp/proto/usr/bin/blah` appear in the image at
+`/usr/bin/blah`
```
Comment (citrus-it) [inline docs/how-to-run-bench-gimlet.adoc:65]: Is it worth giving more specific details about what you use this for? For example, are you jumping ahead to building the `-m non-gimlet` sled-agent, then dropping that and its manifest into the additional proto before coming back and building this image? If the `-P` is omitted you'll get a clean image that you'll need to put sled agent back on each time it reboots - maybe that's ok for developing in a bench setup?

File: docs/how-to-run-bench-gimlet.adoc [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+./helios-build experiment-image \
+  -N "gimlet-sn21-$ds" \
+  -p helios-netdev=https://pkg.oxide.computer/helios-netdev \
+  -o "/net/catacomb/data/staff/dock/gimlet-sn21/os-$ds" \
+  -F optever=0.23 \
+  -B \
+  -P alanproto
+----
+
+The `-F optever=0.23` is subject to change.
```
Comment (citrus-it) [inline docs/how-to-run-bench-gimlet.adoc:82]: Perhaps suggest checking `tools/opte_version` in this repo for the latest version to use?

File: docs/how-to-run-bench-gimlet.adoc [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+Once this image is built, you must install it on your gimlet.
```
Comment (karencfv) [inline docs/how-to-run-bench-gimlet.adoc:83]: I'll write documentation on this process on Monday and link it here after I've created the doc :)

**Comment by leftwo** [comment]
Everyone, thanks for all the help and suggestions on this document!

I'm going to roll in what has been suggested so far, then merge this up to main so we can all continue to
iterate on it, while opening up visibility to more people.

File: docs/how-to-run-bench-gimlet.adoc [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+----
+[[internal_services_ip_pool_ranges]]
+first = "172.20.24.10"          <--- first address of pool used for internal services (like Nexus)
+last = "172.20.24.15"           <--- last address of pool used for internal services
+
+# Configuration to bring up boundary services and make nexus reachable from the outside
+[rack_network_config]
+gateway_ip = "172.20.24.1"    <--- ip address of your default gateway / router
+infra_ip_first = "172.120.24.20" <--- first address of pool used for rack networking
+infra_ip_last = "172.20.24.50"  <--- last address of pool used for rack networking
```
[the remark this answers is not in the document]
Comment (leftwo) [reply docs/how-to-run-bench-gimlet.adoc]: Done

File: docs/how-to-run-bench-gimlet.adoc [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+----
+
+=== Gimet non-gimlet configuration changes
+
+These changes are how we tell Nexus/Omicron what I.P. addresses it will use
+for internal services that require external network access, and what addresses
+we use for ingress and egress of network traffic from virtual machines we
+create.
+
+Update these two sections of `smf/seld-agent/non-gimlet/config-rss.toml` with
```
[the remark this answers is not in the document]
Comment (leftwo) [reply docs/how-to-run-bench-gimlet.adoc]: done!

File: docs/how-to-run-bench-gimlet.adoc (hunk 5 above)
Comment (leftwo) [reply docs/how-to-run-bench-gimlet.adoc:65]: The "clean image that needs sled-agent" is the way I usually roll, but, yeah, people might want to do it
any number of ways.  I don't know exactly how to drop Omicron on top of the image either, so I did not
want to include instructions as I would not get them correct :)

I'll see if I can re-word this a little to get across your point though.

**Review by leftwo — COMMENTED** [review]
Hmm, I think I've mis-merged this with main and now it's showing all sorts of other
stuff that I don't think is part of this change.  I may just take the one file this PR 
actually changed and make another PR out of it.  I'll reference this one so PR comments
will have a history somewhere.

File: docs/how-to-run-bench-gimlet.adoc [hunk 10]
```diff
@@ -0,0 +1,324 @@
...
+  -p helios-dev=https://pkg.oxide.computer/helios/2/dev \
+  -o "/net/catacomb/data/staff/dock/gimlet-sn21/os-$ds" \
+  -F optever=0.23 \
+  -B \
+  -P alanproto
+----
+
+The `-F optever=0.23` is subject to change.  You can discover the current
+Omicron required version by looking in the Omicron repo in `tools/opte_version`
+Once this image is built, you must install it on your gimlet.
```
Comment (karencfv) [inline docs/how-to-run-bench-gimlet.adoc:98]:
```suggestion
Once this image is built, you must https://github.com/oxidecomputer/meta/blob/master/engineering/lab/gimlet.adoc#installing-or-updating-helios[install it on your gimlet].
```

**Comment by jordanhendricks** [comment]
Any reason we can't merge this soon? I find myself linking people to this PR a lot, so it certainly has some use. We could put a big warning at the top if there are paper cuts or incomplete parts.

**Comment by leftwo** [comment]
> Any reason we can't merge this soon? I find myself linking people to this PR a lot, so it certainly has some use. We could put a big warning at the top if there are paper cuts or incomplete parts.

I was worried this had now become too out of date to be useful, and I think that is true.
I'm going to abandon this PR, but keep the file and figure out what useful parts we can stuff elsewhere.

**Pull request closed without merging**